### DISCUSSION TITLE
cowboy 2.9 - read_urlencoded_body/1 vs read_body/1

### DISCUSSION THREAD

**Discussion by ferencboi** [ninenines/cowboy#1573] [Q&A]
Hey there,
I have noticed when I am sending a POST request with payload (JSON), read_body/1 does not return the JSON string the client has sent via a jQuery ajax POST request.
However read_urlencoded_body/1 is able to read the payload and returns a single tuple list containing the escaped JSON string and the boolean 'true'.
Is this the expected behavior? Shouldn't read_body return the JSON string instead? Maybe I am missing a client side request parameter?

**Comment by essen** [comment]
Please show the code.

**Comment by ferencboi** [comment]
Ok, found the issue.
In the single router I currently have I start with **{ok, PostVals, Req} = cowboy_req:read_urlencoded_body(Req0)**, which will change the state of Req0, so when in the next line I want to get the payload with **read_body/1**, read_body/1 yields <<>> even if I use Req0 or Req as argument.
The problem is with my implementation. I have to pre-check the Content-Type header and process the POST data accordingly.
